### PR TITLE
Change the behavior of `GetNodalFESpace` to always return a `FiniteElementSpace`.

### PR DESCRIPTION
**Pull request by YohannDudouit**
The current behavior of `Mesh::GetNodalFESpace` is extremely bug prone, and results in relatively hard to track bugs. This PR modifies the behavior of the `GetNodalFESpace` method to never return a `null` pointer, and always return a valid `FiniteElementSpace`.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3450](https://github.com/mfem/mfem/pull/3450) | @YohannDudouit | @pazner | @v-dobrev + @tzanio + @mlstowell | 2/8/23 | ⌛due 2/22/23 | ⌛due 3/1/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

File: mesh/mesh.cpp [hunk 1]
```diff
@@ -5344,7 +5344,8 @@ void Mesh::SetNodalGridFunction(GridFunction *nodes, bool make_owner)
 
 const FiniteElementSpace *Mesh::GetNodalFESpace() const
 {
-   return ((Nodes) ? Nodes->FESpace() : NULL);
+   const_cast< Mesh* >(this)->EnsureNodes();
```
[the remark this answers is not in the document]
Comment (YohannDudouit) [reply mesh/mesh.cpp:5347]: I'll try to take a look to see if I find a simple solution.

[the remark this answers is not in the document]
Comment (pazner) [reply mesh/mesh.cpp:5347]: One simple approach:

```diff
diff --git a/mesh/mesh.cpp b/mesh/mesh.cpp
index 7b0463ac6..78f68b98d 100644
--- a/mesh/mesh.cpp
+++ b/mesh/mesh.cpp
@@ -5344,7 +5344,7 @@ void Mesh::SetNodalGridFunction(GridFunction *nodes, bool make_owner)
 
 const FiniteElementSpace *Mesh::GetNodalFESpace() const
 {
-   const_cast< Mesh* >(this)->EnsureNodes();
+   if (Nodes == nullptr) { const_cast<Mesh*>(this)->EnsureNodes(); }
    return Nodes->FESpace();
 }
 ```

**Comment by tzanio** [comment]
Should we be worried about the possibility that parts of mfem or external codes rely on the `null` pointer, to distinguish linear from high-order meshes?

My guess is: "no", but I want to check.

**Comment by pazner** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by mlstowell** [comment]
On the one hand I agree with @YohannDudouit that returning a NULL pointer can often lead to bugs when users don't check for a NULL. This kind of return value checking was routine in C programming but C++ programmers often don't think that way anymore. On the other hand MFEM has made this design choice and used it in numerous places throughout the code. Is there a particularly compelling reason to change this return value?

Also, creation of a `Nodes` `GridFunction` and its associated `FiniteElementSpace` in a `const` member function seems like an unusually large violation of the `const`-ness of the mesh object.

For these two reasons I must reluctantly argue against this PR.

**Comment by YohannDudouit** [comment]
@mlstowell The reason for this change is that this method's behavior has resulted in many bugs for myself and other MFEM users. I have also helped many MFEM users track bugs resulting from this behavior.

I would also argue that considering a bad practice a design choice is a bit far fetched. Over-using `NULL` was already a bad practice in C, and I would go further by saying that it is admitted today that inventing `nullptr` was a mistake. Speaking at a software conference in 2009, Tony Hoare apologized for inventing the null reference: "I call it my billion-dollar mistake". You may be interested to read these articles to understand why:
https://hinchman-amanda.medium.com/null-pointer-references-the-billion-dollar-mistake-1e616534d485
https://hackernoon.com/null-the-billion-dollar-mistake-8t5z32d6

There are many ways to misuses `nullptr`, the issue here is about functions returning `NULL`. There is an infinite number of better design solutions than returning `NULL`, the most common modern solution being the use of a class similar to [std::optional](https://en.cppreference.com/w/cpp/utility/optional). As an exercise it can be interesting to understand why Rust's [option](https://doc.rust-lang.org/std/option/) is theoretically superior to C++ [std::optional](https://en.cppreference.com/w/cpp/utility/optional).

I also suspect that the "design choice" of not having a `FiniteElementSpace` for the mesh nodes when the mesh is first order comes from another common bad practice: premature optimization. 
> There is no doubt that the grail of efficiency leads to abuse. Programmers waste enormous amounts of time thinking about, or worrying about, the speed of noncritical parts of their programs, and these attempts at efficiency actually have a strong negative impact when debugging and maintenance are considered. We should forget about small efficiencies, say about 97% of the time: premature optimization is the root of all evil.
> Yet we should not pass up our opportunities in that critical 3%. A good programmer will not be lulled into complacency by such reasoning, he will be wise to look carefully at the critical code; but only after that code has been identified. It is often a mistake to make a priori judgements about what parts of a program are really critical, since the universal experience of programmers who have been using measurement tools has been that their intuitive guesses fail. After working with such tools for seven years, I've become convinced that all compilers written from now on should be designed to provide all programmers with feedback indicating what parts of their programs are costing the most; indeed, this feedback should be supplied automatically unless it has been specifically turned off.

I do not believe that __always__ having a `FiniteElementSpace` for the mesh nodes would result in any substantial performance impact, and as stated above this kind of optimization should be based on measured performance rather than intuition. I think this "optimization" should be opt-in as it results in unsafe and bug prone interface.

If the MFEM project had explicit practices regarding the use of pointers vs references, the convention that anything that is a pointer __must__ be checked for being potentially `NULL` could make MFEM more user-friendly. However, one should be careful that deducing and exploiting the semantic of `NULL` in a given context is not always possible. Too many times I had to re-design code because an MFEM method was returning `NULL`, and I couldn't do anything useful from that information ( I'm thinking of `GetIntegrationRule` for instance which almost always returns `NULL` ).

Since the use of pointers/references is arbitrary in MFEM, if we want to keep the existing behavior I would recommend deprecating the current function and renaming it to `MaybeGetNodelFESpace` as a hint to users that the method frequently returns `NULL`.

**Comment by mlstowell** [comment]
@YohannDudouit, you make many good points but the solution proposed in this PR will break other user's code. Consider someone whose code uses `Mesh::GetVertex` to access the coordinates of the first order nodes directly. After calling, perhaps indirectly, your proposed `const` method `Mesh::GetNodalFESpace()` their vertices are suddenly and inexplicably abandoned. The vertex coordinate arrays still contain data but that data will no longer be updated if the mesh geometry changes. That is going to be a much harder bug to track down than unexpectedly accessing a NULL pointer.

Again, I don't disagree with you on most of your points. However, I think we have painted ourselves into this particular corner and we cannot easily get out of it without making a bigger mesh.

**Comment by sebastiangrimberg** [comment]
Perhaps this is a breaking change worthwhile considering for v5.0? I agree with @YohannDudouit and it seems like just always using a `FiniteElementSpace` would lead to a simpler and more maintainable `Mesh` type. Just chiming in my opinion as I came across this issue.

**Comment by pazner** [comment]
Given @mlstowell's comments, this is liable to be a breaking change in actual user code (though I do agree with @YohannDudouit that we want to move in the direction of less bug-prone code, using fewer null pointers, etc.). I think we should postpone to MFEM 5.0. I'll pause the review on this temporarily, and we can bring it back under consideration when we start working towards MFEM 5.0.

**Comment by kmittal2** [comment]
Late to the party.. if we do not want to introduce something that could hurt unsuspecting users, why can't we abort when we hit this condition? cc: @YohannDudouit @pazner 
I know other external users who have this bug and it is quite painful to track such segfaults. At the very least, we could add an assertion to catch this in debug mode, although I still think it is better to abort (if it breaks someones code, they will simply need to check if `Nodes` exist instead of relying on Null coming from this method).

**Pull request closed without merging**